PLT-456: Propagate SrcSpans to UPLC

zliu41 opened the pull request:
`plc` can now compile into `CompiledCodeDebug`. Also added `compileDebug`.

One bit of concern is that, as seen in the `plc.golden` files, `-g` gives us some SrcSpans, but many are still missing. Also for some reason the same `Term` gets overlapping SrcSpans (which I guess we can simply merge).

michaelpj commented on plutus-tx-plugin/src/PlutusTx/PIRTypes.hs:
```diff
@@ -2,11 +2,13 @@ module PlutusTx.PIRTypes where
 type PIRType uni = PIR.Type PIR.TyName uni Ann
 type PIRTerm uni fun = PIR.Term PIR.TyName PIR.Name uni fun Ann
 type PIRProgram uni fun = PIR.Program PIR.TyName PIR.Name uni fun ()
+type PIRProgramDebug uni fun = PIR.Program PIR.TyName PIR.Name uni fun SrcSpans
```
So... why not just do this always? I mean, does it actually _cost_ us anything noticeable to put these in?

michaelpj replied:
This seems like the key design decision. A lot of this PR would be simpler without it!

zliu41 replied:
In production we need `PIRProgram`, not `PIRProgramDebug`.
This implementation converts the latter to the former basically already at the very end (in `runCompiler`), so I don't see how it can be made simpler.

michaelpj replied:
I don't see where the problem comes up. We already throw away annotations when serialising, this just means that the annotations we throw away are bigger. Or to put it another way, `runCompiler` is not "the very end", that's "where `CompiledCode` gets used", and I claim that already handles this.

zliu41 replied:
> We already throw away annotations when serialising

Does that mean we didn't need `CompiledCodeDebug` in the first place?

michaelpj replied:
Quite probably. I don't have it _all_ loaded into my head, but my current expectation is that it would all work fine if you just changed the annotation type unilaterally.

zliu41 replied:
Ok. I'll think about it. I think the main issue right now is that we are not getting enough `SrcSpans`. Suggestions are welcome.

michaelpj replied:
Poke around in GHC or ask in #ghc is probably the easiest way. "Is there a way to get GHC to add src ticks for every sub-expression" is basically what you're asking, I think.

zliu41 replied:
@michaelpj Throwing away annotations when serialising means `decode . encode /= id`. Is that a problem?

michaelpj reviewed (commented):
Haven't reviewed properly, just dropping in what seems to be the most important question.

zliu41 commented:
@michaelpj As you suggested, I removed `CompiledCodeDebug`, and use `SrcSpans` for annotation directly in `CompiledCode`.

Also:
- Removed the `Flat (CompiledCodeIn uni fun ann a)` instance. It is not actually used anywhere, except in `BasicValidators.hs` under `doc`.
- `serialiseCompiledCode` discards the annotations.
- One thing slightly awkward is `applyCode`. It's not totally clear how to handle the annotations.

We will also most likely need to implement some heuristics to post-process the `SrcSpan`s, because as it stands, some `CoreExpr`s have no `SrcSpan` while some others have multiple. GHC's backend itself uses some heuristics ([ref1](https://downloads.haskell.org/~ghc/9.2.4/docs/html/users_guide/debug-info.html#ghc-flag--g), [ref2](https://www.haskell.org/ghc/blog/20200404-dwarf-2.html)). Another option is to implement another plugin to capture the `SrcSpans` from the parser. Either way, that's an orthogonal problem and will be tackled separately.

michaelpj commented on plutus-tx/src/PlutusTx/Code.hs:
```diff
@@ -111,18 +108,23 @@ instance HasErrorCode ImpossibleDeserialisationFailure where
+getPlcNoAnn
+    :: (PLC.Closed uni, uni `PLC.Everywhere` Flat, Flat fun)
+    => CompiledCodeIn uni fun a -> UPLC.Program UPLC.NamedDeBruijn uni fun ()
+getPlcNoAnn = (() <$) . getPlc
```
```suggestion
getPlcNoAnn = void . getPlc
```

michaelpj commented on plutus-tx/src/PlutusTx/Code.hs:
```diff
@@ -131,7 +133,12 @@ getPir wrapper = case wrapper of
+getPirNoAnn
+    :: (PLC.Closed uni, uni `PLC.Everywhere` Flat, Flat fun)
+    => CompiledCodeIn uni fun a -> Maybe (PIR.Program PIR.TyName PIR.Name uni fun ())
+getPirNoAnn = fmap (() <$) . getPir
```
```suggestion
getPirNoAnn = fmap void . getPir
```

zliu41 commented:
> ... with what goal? I'm not sure what you're trying to achieve.

Ideally we want each `CoreExpr` to have a single `SrcSpan` (or multiple non-overlapping ones with CSE), so that we can show source locations while stepping through Plutus Core. But that's not the case. So the heuristics could be something like
- If a `Lam` does not have `SrcSpan` but its body does, use the body's.
- If a `CoreExpr` have multiple `SrcSpan`s, discard the ones that contain other `SrcSpan`s.

michaelpj commented:
Okay, that makes sense. Can be a Note. 

Probably should discuss this elsewhere, but just dumping some thoughts. We have two problems:
1. Nodes missing srcspans
2. Nodes with multiple srcspans (not sure how often this happens?)

For 1 I would have thought we could just use the nearest enclosing srcspan in most cases? Unclear whether we want to support missing srcspans in the evaluator or whether we want to insist that we're given a program that has them everywhere.

For 2 I think your idea of discarding subsumed srcspans makes sense (although maybe the more precise one is better? unclear). I wonder how often this happens in practice, or whether we can get multiple locations where neither subsumes the other. Again, maybe we want to support multiple scrspans in the evaluator for generality, and then we can try and tidy them up beforehand. Or we could insist on exactly one.

michaelpj commented on plutus-tx-plugin/test/Plugin/Debug/fib.plc.golden:
```diff
@@ -0,0 +1,381 @@
+        (strict)
+        (vardecl
+          fromList []
+          fib
```
`hoistExpr` would be the place.

zliu41 replied:
Nice. Unfortunately it didn't make a difference for these two test cases.

michaelpj replied:
How weird. Surely GHC must put the definition locations of functions _somewhere_!

bezirg replied:
This is strange indeed, if I find time I will try to check manually if fib in the example above contains any SrcSpans

zliu41 commented on plutus-tx-plugin/test/Plugin/Primitives/Spec.hs:
```diff
@@ -45,21 +45,21 @@ primitives = testNested "Primitives" [
   , goldenUEval "ifThenElseApply" [ toUPlc ifThenElse, toUPlc int, toUPlc int2 ]
   , goldenPir "emptyByteString" emptyByteString
-  , goldenUEval "emptyByteStringApply" [ getPlc emptyByteString, liftProgram Builtins.emptyByteString ]
+  , goldenUEval "emptyByteStringApply" [ getPlcNoAnn emptyByteString, liftProgram Builtins.emptyByteString ]
```
But the `lift*` functions can't know the source locations, so it would always be `mempty`, right?

michaelpj replied:
Right, or `Default` or we could pass it as an argument... The only real advantage would be that you'd get something out with a default-annotation-of-your-type, rather than having to convert your existing AST to have no annotations. Really not a big deal and unclear if it would even be nicer.

zliu41 replied:
Hmm let's punt on this. I'm still not sure whether every annotation type should necessarily have a `Default` instance.

zliu41 commented on plutus-tx-plugin/src/PlutusTx/Plugin.hs:
```diff
@@ -417,8 +417,9 @@ runCompiler moduleName opts expr = do
-    when (_posDumpPir opts) . liftIO $ dumpFlat spirP "simplified PIR program" (moduleName ++ ".pir-simplified.flat")
+    let spirPNoAnn = PIR.Program () . void $ spirT
+        spirP = PIR.Program mempty . fmap getSrcSpans $ spirT
+    when (_posDumpPir opts) . liftIO $ dumpFlat spirPNoAnn "simplified PIR program" (moduleName ++ ".pir-simplified.flat")
```
Perhaps with a different flag, say "dump-pir-debug".

michaelpj replied:
yeah, I guess this is fine for now since it's the status quo

zliu41 commented on plutus-tx-plugin/src/PlutusTx/Plugin.hs:
```diff
@@ -417,8 +417,9 @@ runCompiler moduleName opts expr = do
-    let spirP = PIR.Program () . void $ spirT
-    when (_posDumpPir opts) . liftIO $ dumpFlat spirP "simplified PIR program" (moduleName ++ ".pir-simplified.flat")
+    let spirPNoAnn = PIR.Program () . void $ spirT
+        spirP = PIR.Program mempty . fmap getSrcSpans $ spirT
```
Good point. I think this is an argument for the original implementation, in which the compiler takes the annotation type as an argument and can generate Plutus Core with different annotations. If the annotation type is hardcoded, then it should probably include the `Provenance`.

zliu41 replied:
A problem is `Provenance` has no default value (it doesn't have a `Monoid` instance).

zliu41 replied:
Since

> this is all going to have to be redone somehow once we do anything serious with the versions

which I totally agree, I think we can fix this at that point.

michaelpj replied:
> I think this is an argument for the original implementation, in which the compiler takes the annotation type as an argument and can generate Plutus Core with different annotations.

Why? I think we should just always produce the maximally informative version. `Provenance SrcSpans` is strictly more informative than just `SrcSpans`, I think, and you can always throw away the `Provenance` stuff.

> A problem is Provenance has no default value (it doesn't have a Monoid instance).

Where do we need that? For `applyProgram` I guess? You could fix that to only need `Semigroup` by `<>` ing the annotations on the versions instead of using `mempty` I guess. That function is already all hacks, this doesn't make it much worse!

michaelpj replied:
> which I totally agree, I think we can fix this at that point.

(this comment got into the wrong thread, but I totally agree, just wanted to register my feelings about `applyProgram` :sweat_smile: )

michaelpj replied:
happy to leave the provenance stuff for later, unless you want to get it out of the way while you're touching this code

zliu41 commented on plutus-tx-plugin/src/PlutusTx/Options.hs:
```diff
@@ -43,7 +43,8 @@ data PluginOptions = PluginOptions
     , _posPedantic                       :: Bool
-    , _posVerbosity                      :: Verbosity
+    , _posVerbose                        :: Bool
+    , _posDebug                          :: Bool
```
Oops, I don't know how that happened.

zliu41 commented on plutus-core/untyped-plutus-core/src/UntypedPlutusCore.hs:
```diff
@@ -22,10 +22,16 @@ import PlutusCore.Default qualified as PLC
 -- | Take one UPLC program and apply it to another.
-applyProgram :: Program name uni fun () -> Program name uni fun () -> Program name uni fun ()
-applyProgram (Program _ _ t1) (Program _ _ t2) = Program () (PLC.defaultVersion ()) (Apply () t1 t2)
+applyProgram ::
```
Yeah, I had to hold my nose when I made this change :joy:

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -454,7 +455,10 @@ hoistExpr var t = do
+    let addSpan = case getVarSourceSpan var of
+            Nothing  -> id
+            Just src -> fmap . fmap . addSrcSpan $ spanToSpan src
+    addSpan $ case maybeDef of
```
This is wrong, I think. The expression here is the expression we're going to insert at the _use site_, i.e. we see `fib` and this will eventually produce a variable reference to `fib`. _That_ shouldn't have a location based on the definition location of `fib`, the _body_ of `fib` should have that. So I think this needs to go on line 472 instead.

michaelpj replied:
472 in the new diff, the call to `compileExpr`, anyway

michaelpj commented on plutus-tx/src/PlutusTx/Code.hs:
```diff
@@ -45,14 +47,39 @@ data SrcSpan = SrcSpan
+        showString "{ "
+            . showString
+                ( case Set.toList xs of
+                    [] -> "no-src-span"
```
I'd have probably just left `{}` but this is fine

bezirg commented on plutus-core/untyped-plutus-core/src/UntypedPlutusCore.hs:
```diff
@@ -1,13 +1,13 @@
-    , PLC.DefaultFun
-    , mkDefaultProg
-    ) where
+    module Export,
+    Term (..),
+    Program (..),
+    applyProgram,
+    parseScoped,
+    PLC.DefaultUni,
+    PLC.DefaultFun,
+    mkDefaultProg,
+) where
```
Why is ths format change? I like more the old one :)

zliu41 replied:
I think I ran fourmolu on it..

zliu41 commented on plutus-tx-plugin/src/PlutusTx/Annotation.hs:
```diff
@@ -38,3 +39,6 @@ annAlwaysInline = Ann{annInline = AlwaysInline, annSrcSpans = mempty}
 -- | Create an `Ann` with `MayInline`.
 annMayInline :: Ann
 annMayInline = Ann{annInline = MayInline, annSrcSpans = mempty}
+
+addSrcSpan :: SrcSpan -> Ann -> Ann
+addSrcSpan s (Ann i (SrcSpans ss)) = Ann i (SrcSpans $ Set.insert s ss)
```
I think this is more convenient because we can write `fmap (addSrcSpan ss) term`.

zliu41 commented on plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:
```diff
@@ -874,14 +878,25 @@ getSourceSpan _ GHC.ProfNote{GHC.profNoteCC=cc} =
+getVarSourceSpan :: GHC.Var -> Maybe GHC.RealSrcSpan
+getVarSourceSpan = GHC.srcSpanToRealSrcSpan . GHC.nameSrcSpan . GHC.varName
+
+spanToSpan :: GHC.RealSrcSpan -> SrcSpan
+spanToSpan sp = SrcSpan
+    { srcSpanFile = GHC.unpackFS (GHC.srcSpanFile sp),
+      srcSpanSLine = GHC.srcSpanStartLine sp,
+      srcSpanSCol = GHC.srcSpanStartCol sp,
+      srcSpanELine = GHC.srcSpanEndLine sp,
+      srcSpanECol = GHC.srcSpanEndCol sp
+    }
```
Sounds good. Updated.

zliu41 commented on plutus-tx/src/PlutusTx/Code.hs:
```diff
@@ -26,18 +26,70 @@ import Flat.Decoder (DecodeException)
 -- we compile newtypes the same as their underlying types, but people probably just
 -- shouldn't coerce the final parameter regardless, so we play it safe with a nominal role.
-type role CompiledCodeIn representational representational representational nominal
+type role CompiledCodeIn representational representational nominal
```
@bezirg see this thread: https://github.com/input-output-hk/plutus/pull/5008#discussion_r1047424166
